src/main/java/org/opensearch/knn/index/vectorvalues/KNNMergeVectorValues.java

```diff
@@ -0,0 +1,120 @@
+        }
+        final List<KNNVectorValuesSub<FloatVectorValues>> subs = new ArrayList<>();
+        for (int i = 0; i < mergeState.knnVectorsReaders.length; i++) {
+            KnnVectorsReader knnVectorsReader = mergeState.knnVectorsReaders[i];
+            if (knnVectorsReader != null) {
+                FloatVectorValues values = knnVectorsReader.getFloatVectorValues(fieldInfo.getName());
+                if (values != null) {
+                    final Bits liveDocs = mergeState.liveDocs[i];
+                    final int liveDocsCardinality = cardinality(liveDocs, Math.toIntExact(values.cost()));
+                    subs.add(new KNNVectorValuesSub<>(mergeState.docMaps[i], values, liveDocsCardinality));
+                }
+            }
+        }
+        return new KNNVectorValuesIterator.MergeFloat32VectorValuesIterator(subs, mergeState);
+    }
+
+    /**
+     * Gets list of {@link KNNVectorValuesSub} for {@link ByteVectorValues} from a merge state. This can be further
+     * used to create an iterator for getting the docs and its vector values
+     * @param fieldInfo
+     * @param mergeState
+     * @return List of KNNVectorSub
+     * @throws IOException
+     */
+    public static KNNVectorValuesIterator.MergeByteVectorValuesIterator mergeByteVectorValues(FieldInfo fieldInfo, MergeState mergeState)
+        throws IOException {
+        assert fieldInfo != null && fieldInfo.hasVectorValues();
+        if (fieldInfo.getVectorEncoding() != VectorEncoding.BYTE) {
+            throw new UnsupportedOperationException("Cannot merge vectors encoded as [" + fieldInfo.getVectorEncoding() + "] as BYTE");
+        }
+        final List<KNNVectorValuesSub<ByteVectorValues>> subs = new ArrayList<>();
+        for (int i = 0; i < mergeState.knnVectorsReaders.length; i++) {
+            KnnVectorsReader knnVectorsReader = mergeState.knnVectorsReaders[i];
+            if (knnVectorsReader != null) {
+                ByteVectorValues values = knnVectorsReader.getByteVectorValues(fieldInfo.getName());
+                if (values != null) {
+                    final Bits liveDocs = mergeState.liveDocs[i];
+                    final int liveDocsCardinality = cardinality(liveDocs, Math.toIntExact(values.cost()));
+                    subs.add(new KNNVectorValuesSub<>(mergeState.docMaps[i], values, liveDocsCardinality));
+                }
+            }
+        }
+        return new KNNVectorValuesIterator.MergeByteVectorValuesIterator(subs, mergeState);
+    }
+
+    private static int cardinality(final Bits liveDocs, final int defaultCount) {
+        if (liveDocs == null) {
+            return defaultCount;
+        }
+
+        if (liveDocs instanceof FixedBitSet) {
+            return ((FixedBitSet) liveDocs).cardinality();
+        }
+
+        int count = 0;
+        for (int index = 0; index < liveDocs.length(); index++) {
+            count += liveDocs.get(index) ? 1 : 0;
+        }
```

Comment:
This will iterate over all the bits in the bitset, rather I would suggest VectorValues to iterate over the docIds and then see if the docIds is present in bitset or not. Because it can happen for a field only few docIds are present in liveDocs.

Reply:
That requires another call to `knnVectorsReader.get[Float/Byte]VectorValues` for each segment, which is an anticipated latency driver. IO operation might be more expensive than looping through considering merges

Reply:
Check my comment here: https://github.com/opensearch-project/k-NN/pull/2135#discussion_r1775784103 using live docs will not give you correct doc count.